### ISSUE TITLE
判断条件的依据

### ISSUE THREAD

**Issue by exectrl** [Zefan-Cai/PyramidKV#17]
感谢您分享的代码，请问在pyramidkv_utils.py中的
elif q_len < (self.max_capacity_prompt - self.window_size) * 2:             indices = attn_cache.topk(self.max_capacity_prompt - self.window_size, dim=-1).indices
选用这个判断条件的依据是什么？还有后面的else:            indices = attn_cache.topk(max_capacity_prompt, dim=-1).indices。希望获取作者解答，感谢！

**Comment by exectrl** [comment]
> 在PyramidKV的做法下，最底层的kv cache budget就是接近(self.max_capacity_prompt - self.window_size) * 2。如果q_len小于(self.max_capacity_prompt - self.window_size) * 2，那么在最底层budget会大于q_len，所以就加入了这个判断条件防止报错，在这个条件下动态调整budget不生效。后面else如果生效的话，动态调整budget就会生效

非常感谢您的及时回复！还有一个问题想要请教一下：if key_states.shape[-2] == kv_seq_len:
            self.kv_seq_len = kv_seq_len
            key_states_compress, value_states_compress = self.kv_cluster.update_kv(key_states, query_states, value_states, attention_mask, self.num_key_value_groups)
            past_key_value.update(key_states_compress, value_states_compress, self.layer_idx, cache_kwargs)
        else:
            self.kv_seq_len += q_len
            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)。在上述代码中，为什么if里面的past_key_value.update没有返回值呢？我去看了cache_util.py的源码是有return的啊，这个还挺困惑的。

**Comment by Zefan-Cai** [comment]
> > 在PyramidKV的做法下，最底层的kv cache budget就是接近(self.max_capacity_prompt - self.window_size) * 2。如果q_len小于(self.max_capacity_prompt - self.window_size) * 2，那么在最底层budget会大于q_len，所以就加入了这个判断条件防止报错，在这个条件下动态调整budget不生效。后面else如果生效的话，动态调整budget就会生效
> 
> 非常感谢您的及时回复！还有一个问题想要请教一下：if key_states.shape[-2] == kv_seq_len: self.kv_seq_len = kv_seq_len key_states_compress, value_states_compress = self.kv_cluster.update_kv(key_states, query_states, value_states, attention_mask, self.num_key_value_groups) past_key_value.update(key_states_compress, value_states_compress, self.layer_idx, cache_kwargs) else: self.kv_seq_len += q_len key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)。在上述代码中，为什么if里面的past_key_value.update没有返回值呢？我去看了cache_util.py的源码是有return的啊，这个还挺困惑的。

可能是有返回值的，但是我代码里没有设置变量去接受这个返回值。你就新建一个变量接收返回值就好了。

**Comment by exectrl** [comment]
> > > 在PyramidKV的做法下，最底层的kv cache budget就是接近(self.max_capacity_prompt - self.window_size) * 2。如果q_len小于(self.max_capacity_prompt - self.window_size) * 2，那么在最底层budget会大于q_len，所以就加入了这个判断条件防止报错，在这个条件下动态调整budget不生效。后面else如果生效的话，动态调整budget就会生效
> > 
> > 
> > 非常感谢您的及时回复！还有一个问题想要请教一下：if key_states.shape[-2] == kv_seq_len: self.kv_seq_len = kv_seq_len key_states_compress, value_states_compress = self.kv_cluster.update_kv(key_states, query_states, value_states, attention_mask, self.num_key_value_groups) past_key_value.update(key_states_compress, value_states_compress, self.layer_idx, cache_kwargs) else: self.kv_seq_len += q_len key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)。在上述代码中，为什么if里面的past_key_value.update没有返回值呢？我去看了cache_util.py的源码是有return的啊，这个还挺困惑的。
> 
> 可能是有返回值的，但是我代码里没有设置变量去接受这个返回值。你就新建一个变量接收返回值就好了。

我可能没有表达清楚。如果走if这个分支的话，没有新的返回值的话，那么后面的这部分代码attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)

    if attention_mask is not None:  # no matter the length, we just slice it
        causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
        attn_weights = attn_weights + causal_mask

    # upcast attention to fp32
    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
    attn_weights = nn.functional.dropout(attn_weights, p=self.attention_dropout, training=self.training)
    attn_output = torch.matmul(attn_weights, value_states)中的key_states和value_states用的是key_states = repeat_kv(key_states, self.num_key_value_groups)
    value_states = repeat_kv(value_states, self.num_key_value_groups)这里的key_states和value_states了？那么key_states_compress, value_states_compress = self.kv_cluster.update_kv(key_states, query_states, value_states, attention_mask, self.num_key_value_groups)
            past_key_value.update(key_states_compress, value_states_compress, self.layer_idx, cache_kwargs)这里的作用体现在哪里呢？不太明白这里面的逻辑。
